File: src/commands/git/amend.ts
import * as prompts from "@clack/prompts";
import { Command } from "commander";
import { getShellError } from "../../utils/errors";
import { getCurrentBranch } from "../../utils/git";

export const amend = new Command("amend")
	.description("Stage tracked changes and amend the last commit")
	.option("-p, --push", "Force push after amending")
	.action(async (opts: { push?: boolean }) => {
		prompts.intro("git amend");

		const s = prompts.spinner();
		try {
			const status = await Bun.$`git status --porcelain`.quiet().text();
			const trackedChanges = status
				.split("\n")
				.filter((l) => l.trim() && !l.startsWith("??"));

			if (trackedChanges.length === 0) {
				prompts.log.warn("No tracked changes to amend.");
				process.exit(0);
			}

			s.start("Staging and amending...");
			await Bun.$`git add -u`.quiet();
			await Bun.$`git commit --amend --no-edit`.quiet();
			s.stop("Amended");

			if (opts.push) {
				const currentBranch = await getCurrentBranch();
				s.start(`Force pushing ${currentBranch}...`);
				await Bun.$`git push --force-with-lease origin ${currentBranch}`.quiet();
				s.stop("Pushed");
			}

			prompts.outro("Done!");
		} catch (e: unknown) {
			s.stop("Failed");
			prompts.log.error(getShellError(e));
			process.exit(1);
		}
	});

File: src/commands/git/amend.rs
use crate::utils::git::{Shell, get_current_branch};
use anyhow::Result;

fn run_cmd(args: &[&str]) -> Result<String> {
    let output = std::process::Command::new("git").args(args).output()?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow::anyhow!("git failed: {stderr}"));
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn execute_amend(shell: &dyn Shell, push: bool) -> Result<()> {
    let status = run_cmd(&["status", "--porcelain"])?;
    let tracked: Vec<&str> = status
        .lines()
        .filter(|l| !l.trim().is_empty() && !l.starts_with("??"))
        .collect();

    if tracked.is_empty() {
        eprintln!("No tracked changes to amend.");
        return Ok(());
    }

    println!("Staging and amending...");
    run_cmd(&["add", "-u"])?;
    run_cmd(&["commit", "--amend", "--no-edit"])?;
    println!("Amended");

    if push {
        let current = get_current_branch(shell)?;
        println!("Force pushing {current}...");
        run_cmd(&["push", "--force-with-lease", "origin", &current])?;
        println!("Pushed");
    }

    println!("Done!");
    Ok(())
}
